### PR TITLE
add tag to shared object

### PR DESCRIPTION
**Pull request by hgarrereyn**
This PR modifies the shared object handle to store a `tag` with cached objects. By default, `shared.acquire` will keep a reference until it is garbage collected. By specifying a tag, the user can force cache eviction if the tag passed with acquire doesn't match the tag stored with the cached object.

This PR does not change public APIs but it is a prerequisite for the model streaming api where `RunInference` may need to force cache eviction of models loaded in memory.

Currently, `shared.py` is migrating to Beam core however `RunInference` still depends on the tfx-bsl version. Eventually this PR will also need to be reflected in the Beam version of `shared.py`.

### CODE REVIEW

**Comment by hgarrereyn** [comment]
@rose-rong-liu  @SherylLuo @aaltay

**Comment by aaltay** [comment]
LGTM. Nice idea.

File: tfx_bsl/beam/shared_test.py [hunk 1]
```diff
@@ -237,6 +237,28 @@ def dummy_acquire_fn():
...
+    shared1 = shared.Shared()
+    shared2 = shared.Shared()
+
+    def acquire_fn_1():
+      return NamedObject('obj_1')
+
+    def acquire_fn_2():
+      return NamedObject('obj_2')
+
+    # with no tag, shared handle does not know when to evict objects
```
Comment (rose-rong-liu) [inline tfx_bsl/beam/shared_test.py:250]: s/with/With

File: tfx_bsl/beam/shared_test.py [hunk 2]
```diff
@@ -237,6 +237,28 @@ def dummy_acquire_fn():
...
+    p1 = shared1.acquire(acquire_fn_1)
+    assert p1.get_name() == 'obj_1'
+    p2 = shared1.acquire(acquire_fn_2)
+    assert p2.get_name() == 'obj_1'
+
+    # cache eviction can be forced by specifying different tags
```
Comment (rose-rong-liu) [inline tfx_bsl/beam/shared_test.py:256]: s/cache/Caches

**Comment by hgarrereyn** [comment]
this was merged

**Pull request closed without merging**